synapse/storage/databases/main/registration.py

```diff
@@ -1042,6 +1052,54 @@ async def update_access_token_last_validated(self, token_id: int) -> None:
             desc="update_access_token_last_validated",
         )
 
+    async def lookup_refresh_token(
+        self, token: str
+    ) -> Optional[RefreshTokenLookupResult]:
+        """Lookup a refresh token with hints about its validity."""
+
+        def _lookup_refresh_token_txn(txn) -> Optional[RefreshTokenLookupResult]:
+            txn.execute(
+                """
+                SELECT
+                    rt.id token_id,
+                    rt.user_id,
+                    rt.device_id,
+                    rt.next_token_id,
+                    (nrt.next_token_id IS NOT NULL) has_next_refresh_token_been_refreshed,
+                    (at.last_validated IS NOT NULL) has_next_access_token_been_used
```

Comment:
I don't think that `at.last_validated` does what you need here (it records the last time that the token was revalidated with user-interactive auth).

The nearest to what you want is probably to join against `user_ips` (by user_id and access_token). `user_ips` is updated in batches so it lags by up to 5 seconds, but hopefully that's acceptable?

Reply:
Done in https://github.com/matrix-org/synapse/pull/9450/commits/262d1ab17405eecba65e045ae75b3f958bb4b599